agent/dns.go

```diff
@@ -764,26 +764,42 @@ func (d *DNSServer) dispatch(remoteAddr net.Addr, req, resp *dns.Msg, maxRecursi
 			MaxRecursionLevel: maxRecursionLevel,
 			EnterpriseMeta:    locality.EnterpriseMeta,
 		}
-		// Support RFC 2782 style syntax
-		if n == 2 && strings.HasPrefix(queryParts[1], "_") && strings.HasPrefix(queryParts[0], "_") {
 
-			// Grab the tag since we make nuke it if it's tcp
-			tag := queryParts[1][1:]
+		tag := ""
+
+		// If the first two query parts begin with _, this is an RFC 2782 style SRV lookup.
+		if strings.HasPrefix(queryParts[0], "_") && strings.HasPrefix(queryParts[1], "_") && n >= 2 {
+			// Support RFC 2782 style syntax in one of the following forms:
+			// 1) _<service-name>._<tag or tcp>[.service][.<datacenter>].<domain>    (n == 2)
```

Comment:
Is [.service] actually optional? The docs say yes ... but I want to double-check that in the code.

Reply:
Confirmed that [.service] is optional for this first RFC 2782 style lookup format.